### PR TITLE
[Naitei] ft:Add_Product_Admin

### PR DESCRIPTION
**Pull request by ThaiKhacDuoc**
## Related Tickets
- ticket redmine
https://edu-redmine.sun-asterisk.vn/issues/78419
## WHAT (optional)
- Change number items `completed/total` in admin page.

## HOW
- I edit js file, inject not_vary_normal items in calculate function.

## WHY (optional)
- Because in previous version - number just depends on `normal` items. But in new version, we have `state` and `confirm_state` depends on both `normal` + `not_normal` items.

## Evidence (Screenshot or Video)

image

image

### CODE REVIEW

File: app/models/product.rb [hunk 1]
```diff
@@ -1,8 +1,23 @@
...
+  validates :description, presence: true
+  validates :price, presence: true, numericality: {greater_than: 0}
+  validates :remain_quantity, presence: true,
+                              numericality: {only_integer: true,
+                                             greater_than_or_equal_to: 0}
+  validates :category_id,
+            presence: true
+  validates :image, attached: true,
+                    content_type: ["image/png", "image/jpg", "image/jpeg"],
+                    size: {less_than: 5.megabytes}
```
Comment (ghost) [inline app/models/product.rb]: 5 với 0 đưa vào setting nhé

File: app/views/admin/products/_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<%= form_with model: [:admin, @product], url: admin_products_path do |f| %>
+  <%= render "shared/error_messages", object: f.object%>
```
Comment (ghost) [inline app/views/admin/products/_form.html.erb]:
```suggestion
  <%= render "shared/error_messages", object: f.object %>
```

File: app/views/admin/products/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<% provide :title, t("admin.products.new.title") %>
+<h1><%= t("admin.products.new.title") %></h1>
```
Comment (ghost) [inline app/views/admin/products/new.html.erb]: bỏ bớt () đi em

**Review by ghost — COMMENTED** [review]
cmt

**Comment by ThaiKhacDuoc** [comment]
ready

File: app/models/category.rb [hunk 4]
```diff
@@ -13,4 +13,6 @@ class Category < ApplicationRecord
                        resize_to_limit: [Settings.digit_160, Settings.digit_160]
   end
   scope :oldest, ->{order(parent_category_id: :asc, created_at: :asc)}
+  has_one_attached :image
+  scope :all_categories, ->{all}
```
Comment (ghost) [inline app/models/category.rb]: Ủa làm như này để làm gì đấy

File: app/controllers/admin/products_controller.rb [hunk 5]
```diff
@@ -3,4 +3,27 @@ class Admin::ProductsController < AdminController
   def index
     @pagy, @products = pagy(Product.newest, items: Settings.page_size)
   end
+
+  def new
+    @product = Product.new
+  end
+
+  def create
+    @product = Product.new(product_params)
```
Comment (ghost) [inline app/controllers/admin/products_controller.rb]: bỏ bớt () đi em

File: app/models/product.rb [hunk 6]
```diff
@@ -1,8 +1,26 @@
 class Product < ApplicationRecord
   acts_as_paranoid
+  PERMITTED_ATTRIBUTES = [:name, :description, :price,
+                          :remain_quantity, :image, :category_id].freeze
```
Comment (ghost) [inline app/models/product.rb]: Dùng %i

**Review by ghost — COMMENTED** [review]
cmt

**Comment by ThaiKhacDuoc** [comment]
ready

File: app/models/product.rb [hunk 7]
```diff
@@ -10,4 +13,19 @@ class Product < ApplicationRecord
...
+                    numericality: {greater_than: Settings.digit_0}
+  validates :remain_quantity, presence: true,
+                              numericality: {
+                                only_integer: true,
+                                greater_than_or_equal_to: Settings.digit_0
+                              }
+  validates :category_id,
+            presence: true
+  validates :image, attached: true,
+                    content_type: ["image/png", "image/jpg", "image/jpeg"],
```
Comment (ghost) [inline app/models/product.rb]: đưa ["image/png", "image/jpg", "image/jpeg"] vào settings nhé

**Review by ghost — COMMENTED** [review]
fix nốt rồi anh merge

**Comment by ThaiKhacDuoc** [comment]
ready

**Pull request merged**